src/docs-assembler/Cli/RepositoryCommands.cs

```diff
@@ -0,0 +1,94 @@
+// Licensed to Elasticsearch B.V under one or more agreements.
+// Elasticsearch B.V licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information
+
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using ConsoleAppFramework;
+using Elastic.Markdown.IO;
+using Microsoft.Extensions.Logging;
+using ProcNet;
+using ProcNet.Std;
+
+namespace Documentation.Assembler.Cli;
+
+public class ConsoleLineHandler(string prefix) : IConsoleLineHandler
+{
+	public void Handle(LineOut lineOut) => lineOut.CharsOrString(
+		r => Console.Write(prefix + ": " + r),
+		l => Console.WriteLine(prefix + ": " + l));
+
+	public void Handle(Exception e) { }
+}
+
+internal class RepositoryCommands(ILoggerFactory logger)
+{
+	private void AssignOutputLogger()
+	{
+		var log = logger.CreateLogger<Program>();
+		ConsoleApp.Log = msg => log.LogInformation(msg);
+		ConsoleApp.LogError = msg => log.LogError(msg);
+	}
+
+	// would love to use libgit2 so there is no git dependency but
+	// libgit2 is magnitudes slower to clone repositories https://github.com/libgit2/libgit2/issues/4674
+	/// <summary> Clones all repositories </summary>
+	/// <param name="ctx"></param>
+	[Command("clone-all")]
+	public async Task CloneAll(Cancel ctx = default)
+	{
+		var configFile = Path.Combine(Paths.Root.FullName, "src/docs-assembler/conf.yml");
+		var config = AssemblyConfiguration.Deserialize(File.ReadAllText(configFile));
+
+		Console.WriteLine(config.Repositories.Count);
+		var dict = new ConcurrentDictionary<string, Stopwatch>();
+		await Parallel.ForEachAsync(config.Repositories,
+			new ParallelOptions { CancellationToken = ctx, MaxDegreeOfParallelism = Environment.ProcessorCount / 4 }, async (kv, c) =>
+			{
+				await Task.Run(() =>
+				{
+					var name = kv.Key;
+					var repository = kv.Value;
+					var checkoutFolder = Path.Combine(Paths.Root.FullName, $".artifacts/assembly/{name}");
+
+					var sw = Stopwatch.StartNew();
+					dict.AddOrUpdate(name, sw, (_, _) => sw);
+					Console.WriteLine($"Checkout: {name}\t{repository}\t{checkoutFolder}");
+					var branch = repository.Branch ?? "main";
+					var args = new StartArguments(
+						"git", "clone", repository.Origin, checkoutFolder, "--depth", "1"
```

Comment:
What about sparse-checkout? Did you already try that?

Reply:
I followed the 4 clone patterns defined here: https://github.blog/open-source/git/get-up-to-speed-with-partial-clone-and-shallow-clone/

But can't remember the results :)

Will go over it again when we set up our runner.

Reply:
Yup. In our case we can only download the docs folder.